docs/developer/workflows.md

```diff
@@ -14,12 +14,22 @@ Workflows should all be run from the `src/ontology/` directory.
 ---
 
 ## Slurp / migration
-These workflows will determine slurpable / migratable terms. That is, terms that are candidates for integration into Mondo.
+These workflows will determine slurpable / migratable terms. That is, terms that are candidates for integration into 
+Mondo.
+
+#### What determines what is a slurpable / migratable term?
+To be migratable, the term (i) must not already be mapped, (ii) must not be excluded (e.g. not in 
+`reports/%_term_exclusions.txt`), and (iii) must not be deprecated / obsolete. Then, (iv) there are conditions 
+concerning its parents. It must either (a) have no parents, or (b) have no valid parents in Mondo (i.e. all of its 
+parent terms are marked obsolete in Mondo), or (c) all its parents must be mapped, and at least 1 of those parent's 
+mappings must be one of `skos:exactMatch` or `skos:NarrowMatch`.  
```

Comment:
The better way to phrase this I think is that:

```
Any existing parents must be either mapped, or deprecated.
```

This implies the special cases you list there:
- no parents
- parents obsolete
- parents mapped

The way you phrase it, you are missing the case where 1 parent is obsolete, another mapped (neither (a), (b), (c) applies in that case).

Reply:
I think you may be onto something, I thought about both the logic and phrasing a few times and it still feels clunky.

However, can you double check that I've implemented correctly? What you are suggesting now is different from waht I was under the impression you were asking for in the past:
(i) the term itself can't be deprecated/obsolete in order to be slurped
(ii) there is no check to see if any of the term's parents are marked obsolete within their ontology
(iii) the term's parents, when checking if they are "qualified parents", it should not be considered qualified if that parent is mapped to a Mondo term which is obsoleted.

This is implemented by looking for the parent term in under `object_id` in `mondo.ssom.tsv` and then checking `row['subject_id'].startswith('obsolete')`.

The `parents_obsolete` column in the [slurp integrity spreadsheet](https://docs.google.com/spreadsheets/d/1kc_RGzdRye7nnCLKYpG0DBKTBQxgezd_0MjJell1Q2c/edit#gid=1042657882) would be more appropriately called `parent_mondo_mappings_obsolete`. Maybe it might be good to look at some of those entries in the spreadsheet to help determine if this logic is currently codified correctly, or if it's even needed at all, or otherwise needs to be somehow changed.

Reply:
Column "L" should be more like `all_parents_mapped_or_obsolete_in_mondo`, else LGTM.

I would say though that the logic is now expressed in a very complex manner. Everything seems right to me apart from the "valid parents" condition. This should IMO look more like:

```
def _valid_parent(parents: List)->Bool:
    for p in parents:
        if !is_mapped(p) and !is_deprecated_in_mondo(p):
             return False
     return True
```

Reply:
Hmm, actually if "all parents mapped", that also implies "all parents mapped or obsolete in Mondo", for in order for the parent to be obsolete in Mondo, it must be mapped in Mondo. Therefore, this would be logically equivalent to any of the below:


```py
def _valid_parents(parents: List) -> Bool:
    for p in parents:
        if !is_mapped(p):
             return False
     return True

valid_parents = not any([not is_mapped(p) for p in parents])

valid_parents = all([is_mapped(p) for p in parents])
```

Reply:
> for in order for the parent to be obsolete in Mondo, it must be mapped in Mondo

I agree with this! I only say that because the mappings from obsolete terms may not show up in your pipeline? If they do, then I agree with you!

Just think of it like this (integrity check of your approach):

**The migration (slurp) table may only ever be empty if and only if all terms from the source are either mapped, deprecated or excluded.** (note the absence of the `parent` condition, which is only relevant for the _order_ in which migration happens).

This should ideally not wait till Thursday (urgent requirement for the paper), just solve it somehow now and we discuss it later again. Also I am OOO next week, but will respond to slack.

Reply:
Hmm, alright. I just added a parameter called `parent_conditions_on`, defaulting to False, and implemented a 1 line solution. If `parent_conditions_on == False`, then it skips all of that logic.

The end result is that the terms in `slurp/%.tsv` will be exactly the same as `reports/%_unmapped_terms.tsv`, which is the same as the list of terms in `reports/%_mapping_status.tsv` where `is_mapped`, `is_deprecated`, and `is_obsolete` are False. I just ran a test and confirmed that was the case.

Reply:
@matentzn When you're done looking at this you might want to hit "Resolve conversation" on this one and continue further discussion [here](https://github.com/monarch-initiative/mondo-ingest/pull/316#discussion_r1199681163).